### ISSUE TITLE
Compiler error: "Call parameter type does not match function signature!"

### ISSUE THREAD

**Issue by Horrowind** [ziglang/zig#3476]
I found similar issues in the bug tracker, but this might be another case. When compiling the following code, the compiler emits an issue:
```zig
const Baz = struct {
    a : i64,
    const baz = Baz { .a = 0 };
};

fn Bar(comptime T : type, n : usize) type {
    return struct {
        e : [n]T,
        pub fn bar() T {
            return T.baz;
        }
    };
}

fn Foo(comptime T : type) type {
    return struct {
        e : Bar(T, 4),
        pub fn foo() Foo(T) {
            return Foo(T) { .e = Bar(T, 4).bar() };
        }
    };
}

pub fn main() anyerror!void {
    const z = Foo(Baz).foo();
}
```

The compiler output is:
```bash
zig build
broken LLVM module found: Call parameter type does not match function signature!
  %1 = getelementptr inbounds %"Foo(Baz)", %"Foo(Baz)"* %0, i32 0, i32 0, !dbg !14271
 %Baz*  call fastcc void @"Bar(Baz,4).bar"(%"Bar(Baz,4)"* sret %1), !dbg !14273

???:?:?: 0x55f4f3d08f3c in ??? (???)


???:?:?: 0x55f4f3ccb917 in ??? (???)


???:?:?: 0x55f4f3c13f23 in ??? (???)


???:?:?: 0x55f4f3c1df3a in ??? (???)


???:?:?: 0x55f4f3bf2f80 in ??? (???)


???:?:?: 0x7eff5b2afbba in ??? (???)


???:?:?: 0x5541d689495640ff in ??? (???)


The following command terminated unexpectedly:
/home/xxx/.local/bin/zig build-exe /tmp/zig_crash/src/main.zig --cache-dir /tmp/zig_crash/zig-cache --name zig_crash --cache on

Build failed. The following command failed:
/tmp/zig_crash/zig-cache/o/h2GExo5iT3nOj9UBaFtN1hIedZA0wg9wwXCXBKzlNc4EGsHJx60AHLVmsco8Y44-/build /home/xxx/.local/bin/zig /tmp/zig_crash /tmp/zig_crash/zig-cache
```
This line is wrong: `return Foo(T) { .e = Bar(T, 4).bar() };`, .e expects something of type `Bar(T, 4)` but it get something of type `T`, i.e. the program is not correct. In some small modifications of the above program the compiler detects this, instead of crashing, yet somehow it does not in this case.

**Comment by rdunnington** [comment]
I'm also seeing this on a build of master commit #1b64a5f5f. Attached is a file you can repro it with. A workaround is to change this line in `fn Slotmap()`:
`next_free: u16`
to
`next_free:usize`

[slotmap.zip](https://github.com/ziglang/zig/files/4023754/slotmap.zip)

(Just figured another data point may help in tracking this bug down). Thanks!

**Comment by Vexu** [comment]
Fixed by 5ea79bfc4a0a1269930d98faee36a8f6cb0b0401

```
./build/a.zig:40:47: error: expected type 'Bar(Baz,4)', found 'Baz'
            return Foo(T) { .e = Bar(T, 4).bar() };
                                              ^
./build/a.zig:28:12: note: Bar(Baz,4) declared here
    return struct {
           ^
./build/a.zig:22:13: note: Baz declared here
const Baz = struct {
            ^
```
